kafka-client/src/main/java/dev/responsive/kafka/internal/db/partitioning/SegmentPartitioner.java

```diff
@@ -0,0 +1,194 @@
+ * <p>
+ * When stream-time for a given kafka partition advances, a tombstone is sent to delete the entire
+ * oldest physical segment (or segments, if it has advanced by more than the segmentInterval).
+ * At the same time, a new physical segment (or segments) will be created and the corresponding
+ * remote partition initialized.
+ * <p>
+ * Let's look at an example case, at a particular moment in time, with the following configuration:
+ * N = 4
+ * retentionPeriod = 100ms
+ * numSegments = 3
+ * segmentInterval = 33
+ * <p>
+ * kafkaPartition | stream-time | minValidTs |    active segment time bounds   |  segmentId ||
+ *           0    |     16      |      0     | 0-32                            | 0          ||
+ *           1    |     88      |      0     | 0-32, 33-65, 66-98              | 0, 1, 2    ||
+ *           2    |     101     |      2     | 0-32, 33-65, 66-98, 99-131      | 0, 1, 2, 3 ||
+ *           3    |     169     |      70    | 66-98, 99-131, 132-164, 165-197 | 1, 2, 3, 4 ||
+ * <p>
+ * NOTE: because we are already dividing up the partition space into segments, we don't further
+ * split things into true sub-partitions based on key. Each kafka partition still maps to multiple
+ * table partitions which should help with the parallelism, but unlike the subpartitioner, the
+ * segment-partitioning scheme is not static and temporal rather than key-based. If data is skewed
+ * in time, this might result in uneven partitions and a need to further subdivide the partition
+ * space.
+ * <p>
+ * For the time being, we simply recommend that users configure the number of segments
+ * similarly to how they would configure the number of sub-partitions for a key-value store.
+ */
+public class SegmentPartitioner implements ResponsivePartitioner<Stamped<Bytes>, SegmentPartition> {
+
+  private final long retentionPeriodMs;
+  private final long segmentIntervalMs;
+
+  public static class SegmentPartition {
+    public final int partitionKey;
+    public final long segmentId;
+
+    public SegmentPartition(final int partitionKey, final long segmentId) {
+      this.partitionKey = partitionKey;
+      this.segmentId = segmentId;
+    }
+  }
+
+  public static SegmentPartitioner create(
+      final ResponsiveWindowParams params
+  ) {
+    final long segmentInterval =
+        StoreUtil.computeSegmentInterval(params.retentionPeriod(), params.numSegments());
+    return new SegmentPartitioner(params.retentionPeriod(), segmentInterval);
+  }
+
+  public SegmentPartitioner(final long retentionPeriodMs, final long segmentIntervalMs) {
+    this.retentionPeriodMs = retentionPeriodMs;
+    this.segmentIntervalMs = segmentIntervalMs;
+  }
+
+  @Override
+  public SegmentPartition tablePartition(final int kafkaPartition, final Stamped<Bytes> key) {
+    // We don't subdivide the kafka partition any further since it's assumed there are
+    // sufficient segments to get good parallelism, but we could make both configurable
```

Comment:
note that it isn't about parallelism and more about bloom filter efficiency, which are only done the the partition key

Reply:
Fair enough...I'm actually going to remove this comment anyways since this is already noted in the main docs section above. It was more a throwaway note than a suggestion to actually implement sub-partitioning within segmented tables 🤷‍♀️